src/cgt/src/ClockGating.cpp

```diff
@@ -0,0 +1,1479 @@
+            << "\"[shape=triangle, color=coral];\n";
+      }
+    }
+    out << "}\n";
+  } else if (auto cell = network->cell(inst)) {
+    out << "subgraph \"" << network->name(inst) << "\" {\n";
+    out << "cluster=true;\n";
+    out << "label=\"" << network->name(cell) << '\n'
+        << network->name(inst) << "\";\n";
+    auto pin_iter = network->pinIterator(inst);
+    while (pin_iter->hasNext()) {
+      auto pin = pin_iter->next();
+      out << '"' << network->name(pin) << "\";\n";
+    }
+  }
+  auto pin_iter = network->pinIterator(inst);
+  while (pin_iter->hasNext()) {
+    auto pin = pin_iter->next();
+    out << '"' << network->name(pin) << "\" [label=\"" << network->portName(pin)
+        << "\"];\n";
+  }
+  auto net_iter = network->netIterator(inst);
+  while (net_iter->hasNext()) {
+    auto net = net_iter->next();
+    out << '"' << network->name(net) << "\";\n";
+    sta::NetPinIterator* pin_iter = network->pinIterator(net);
+    while (pin_iter->hasNext()) {
+      auto pin = pin_iter->next();
+      if (network->direction(pin)->isAnyInput()) {
+        out << '"' << network->name(net) << "\" -> \"" << network->name(pin)
+            << "\";\n";
+      }
+      if (network->direction(pin)->isAnyOutput()) {
+        out << '"' << network->name(pin) << "\" -> \"" << network->name(net)
+            << "\";\n";
+      }
+    }
+  }
+  auto child_iter = network->childIterator(inst);
+  while (child_iter->hasNext()) {
+    auto child = child_iter->next();
+    dumpGraphviz(network, child, out);
+  }
+  out << "}\n";
+}
+
+// Times the given function `f` and logs the time taken. Aggregates the time in
+// microseconds.
+template <typename F>
+static auto timed(Logger* logger, size_t& total_us, const char* label, F f)
+{
+  auto start = std::chrono::steady_clock::now();
+  const auto end_timer = [&] {
+    auto end = std::chrono::steady_clock::now() - start;
+    auto time_us
+        = std::chrono::duration_cast<std::chrono::microseconds>(end).count();
+    debugPrint(
+        logger, utl::CGT, "clock_gating", 2, "{} time: {} μs", label, time_us);
+    total_us += time_us;
+  };
+  if constexpr (std::is_void_v<decltype(f())>) {
+    f();
+    end_timer();
+  } else {
+    auto result = f();
+    end_timer();
+    return std::move(result);
+  }
+}
+
+// Returns the output pin of a register cell, if it exists.
+static sta::Pin* getRegOutPin(sta::dbNetwork* network, sta::Instance* inst)
+{
+  auto cell = network->libertyCell(inst);
+  if (!cell) {
+    return nullptr;
+  }
+  for (auto seq : cell->sequentials()) {
+    if (seq->isRegister() && seq->output()) {
+      sta::LibertyCellPortIterator port_iter(cell);
+      while (port_iter.hasNext()) {
+        auto port = port_iter.next();
+        if (port->direction()->isAnyOutput()) {
+          assert(port->function()->op() == sta::FuncExpr::op_port);
+          assert(port->function()->port() == seq->output());
+          return network->findPin(inst, port);
```

Comment:
A reg can have Q and QN outputs (and scan out).  You need to actually check that the output port is the one in the function.

I'm not certain if you only have QN if it will appear in output or outputInv.  You should verify this works in that case.

Reply:
You mean I should check it as a condition, not assert it (second assert)?

Do `output` and `outputInv` correspond to the two pins in `ff()` in Liberty? Like `ff (IQN,IQNN)`? If so, I don't think I need to care about `outputInv`.